src/java/org/apache/cassandra/index/sai/disk/ByteSliceReader.java

```diff
@@ -0,0 +1,136 @@
+ */
+
+package org.apache.cassandra.index.sai.disk;
+
+import org.apache.lucene.store.DataInput;
+import org.apache.lucene.util.ByteBlockPool;
+
+/* IndexInput that knows how to read the byte slices written
+ * by Posting and PostingVector.  We read the bytes in
+ * each slice until we hit the end of that slice at which
+ * point we read the forwarding address of the next slice
+ * and then jump to it.*/
+final class ByteSliceReader extends DataInput
+{
+    ByteBlockPool pool;
+    int bufferUpto;
+    byte[] buffer;
+    public int upto;
+    int limit;
+    int level;
+    public int bufferOffset;
+
+    public int endIndex;
+
+    public void init(ByteBlockPool pool, int startIndex, int endIndex)
+    {
+
+        assert endIndex - startIndex >= 0 : "startIndex=" + startIndex + " endIndex=" + endIndex;
+        assert startIndex >= 0;
+        assert endIndex >= 0;
+
+        this.pool = pool;
+        this.endIndex = endIndex;
+
+        level = 0;
+        bufferUpto = startIndex / ByteBlockPool.BYTE_BLOCK_SIZE;
+        bufferOffset = bufferUpto * ByteBlockPool.BYTE_BLOCK_SIZE;
+        buffer = pool.buffers[bufferUpto];
+        upto = startIndex & ByteBlockPool.BYTE_BLOCK_MASK;
+
+        final int firstSize = ByteBlockPool.LEVEL_SIZE_ARRAY[0];
+
+        if (startIndex + firstSize >= endIndex)
+        {
+            // There is only this one slice to read
+            limit = endIndex & ByteBlockPool.BYTE_BLOCK_MASK;
+        }
+        else
+            limit = upto + firstSize - 4;
+    }
+
+    public boolean eof()
+    {
+        assert upto + bufferOffset <= endIndex;
+        return upto + bufferOffset == endIndex;
+    }
+
+    @Override
+    public byte readByte()
+    {
+        assert !eof();
+        assert upto <= limit;
+        if (upto == limit)
+            nextSlice();
+        return buffer[upto++];
+    }
+
+    public void nextSlice()
+    {
+
+        // Skip to our next slice
+        final int nextIndex = ((buffer[limit] & 0xff) << 24) + ((buffer[1 + limit] & 0xff) << 16) + ((buffer[2 + limit] & 0xff) << 8) + (buffer[3 + limit] & 0xff);
+
+        level = ByteBlockPool.NEXT_LEVEL_ARRAY[level];
+        final int newSize = ByteBlockPool.LEVEL_SIZE_ARRAY[level];
+
+        bufferUpto = nextIndex / ByteBlockPool.BYTE_BLOCK_SIZE;
+        bufferOffset = bufferUpto * ByteBlockPool.BYTE_BLOCK_SIZE;
+
+        buffer = pool.buffers[bufferUpto];
+        upto = nextIndex & ByteBlockPool.BYTE_BLOCK_MASK;
+
+        if (nextIndex + newSize >= endIndex)
+        {
+            // We are advancing to the final slice
+            assert endIndex - nextIndex > 0;
+            limit = endIndex - bufferOffset;
+        }
+        else
+        {
+            // This is not the final slice (subtract 4 for the
+            // forwarding address at the end of this new slice)
+            limit = upto + newSize - 4;
+        }
+    }
+
+    @Override
+    public void readBytes(byte[] b, int offset, int len)
```

Comment:
Does anything ever call this method? I hunted a little bit for test coverage...but I might not have looked in the right places...

Reply:
I think `readVInt()` only calls `readByte()`. Maybe UnsupportedOperation time :D

Reply:
It's not used by us so I've removed the implementation and replaced it with an `UnsupportedOperationException`.